python/pylibcugraph/pylibcugraph/heterogeneous_biased_neighbor_sample.pyx

```diff
@@ -0,0 +1,424 @@
+    cugraph_sampling_set_dedupe_sources,
+    cugraph_sampling_set_renumber_results,
+    cugraph_sampling_set_compress_per_hop,
+    cugraph_sampling_set_compression_type,
+    cugraph_sampling_set_retain_seeds,
+)
+from pylibcugraph._cugraph_c.sampling_algorithms cimport (
+    cugraph_heterogeneous_biased_neighbor_sample,
+)
+from pylibcugraph.resource_handle cimport (
+    ResourceHandle,
+)
+from pylibcugraph.graphs cimport (
+    _GPUGraph,
+)
+from pylibcugraph.utils cimport (
+    assert_success,
+    assert_CAI_type,
+    assert_AI_type,
+    get_c_type_from_numpy_type,
+)
+from pylibcugraph.internal_types.sampling_result cimport (
+    SamplingResult,
+)
+from pylibcugraph._cugraph_c.random cimport (
+    cugraph_rng_state_t
+)
+from pylibcugraph.random cimport (
+    CuGraphRandomState
+)
+import warnings
+
+# TODO accept cupy/numpy random state in addition to raw seed.
+def heterogeneous_biased_neighbor_sample(ResourceHandle resource_handle,
+                                         _GPUGraph input_graph,
+                                         start_vertex_list,
+                                         start_vertex_offsets,
+                                         h_fan_out,
+                                         num_edge_types,
+                                         bool_t with_replacement,
+                                         bool_t do_expensive_check,
+                                         with_edge_properties=False,
+                                         prior_sources_behavior=None,
+                                         deduplicate_sources=False,
+                                         return_hops=False,
+                                         renumber=False,
+                                         retain_seeds=False,
+                                         compression='COO',
+                                         compress_per_hop=False,
+                                         random_state=None,
+                                         return_dict=False,):
+    """
+    # FIXME: Deprecate biased_neighbor_sample
+    Performs biased neighborhood sampling, which samples nodes from
+    a graph based on the current node's neighbors, with a corresponding fan_out
+    value at each hop. The edges are sampled with biases.
+
+    Parameters
+    ----------
+    resource_handle: ResourceHandle
+        Handle to the underlying device and host resources needed for
+        referencing data and running algorithms.
+
+    input_graph : SGGraph or MGGraph
+        The input graph, for either Single or Multi-GPU operations.
+
+    edge_biases: FIXE: update this - Create edge_biases of type
+        'cugraph_edge_property_view_t' - edge biases not yet supported.
+
+    start_vertex_list: device array type
+        Device array containing the list of starting vertices for sampling.
+
+    start_vertex_offsets: list[int] (Optional)
+        Offsets of each label within the start vertex list.
```

Comment:
> Is this also a device array, or a python list on host? I also noticed the term "label" is used here

Right, we ended up renaming this variable `start_vertex_offsets` to `starting_vertex_label_offsets` in the C++ API to be more descriptive. Updating the PLC API as well

> It also looks like len(start_vertex_offsets) must equal len(start_vertex_list),

It is not equal to the number `start_vertex_list` because `starting_vertex_label_offsets` is in compressed form. below is an example

```
start_vertex_list                                 =    [0, 1, 2, 3, 4, 5, 6]
starting_vertex_label_offsets           =         [0, 3, 5, 7]
start_vertex_labes                             =    [0, 0, 0, 1, 1, 2, 2]           <- computed in the CAPI 
```